### DISCUSSION TITLE
Caching lifecycle scripts

### DISCUSSION THREAD

**Discussion by tgroutars** [vercel/turborepo#1248] [Help] [answered]
Hi,

We have a monorepo that we just migrated from `yarn` to `pnpm`. We're in the process of integrating turborepo too and it looks very promising, super grateful for the maintainers working on this projects 🙏 

The issue we have at the moment is caching lifecycle scripts of our internal packages. Our monorepo contains a bunch of apps, microservices, and libraries. The libraries need to be built for them to work in our apps or other libraries depending on them. We use `prepare` scripts for this (but could also use `postinstall`), so when you `pnpm install` all libraries are being built in the right order. `pnpm` also automatically hard links the libraries into their dependents' node_modules after their lifecycle script has run.

We would like to cache those lifecycle script using turborepo, but I can't find a way to do it properly. I've tried giving those scripts a different name and run them manually after we run `pnpm install`, but on top of not being an ideal devx, it doesn't automatically links the modules again, so the build folder doesn't end up in its dependents' node_modules.

Is there something I'm missing here? I feel like building using lifecycle scripts is a pretty common usecase, but I really don't see how we can cache them using turborepo, and I also can't seem to find any existing discussion or issue about this.

Thanks in advance 🙏

**Comment by nathanhammond** [comment, marked as answer]
So, by default, `turbo` stays out of the package install management lifecycle. That's _deeply_ integrated with each package manager and is not something we can safely intercept. You've noted multiple types of behavior where attempting to cache the results of an install could lead to incorrect behavior. Caching of that lifecycle is best handled by the package managers themselves.

In a monorepo, riding on the `install` rails (`prepare`, `postinstall`) isn't an approach I would recommend; I'd typically just encourage `^build` which will make sure that all things are built correctly. The _first_ `<package manager> install` will set up all the workspace links correctly, and then you'd only call `install` again if you've changed your workspace setup or dependencies.

Then, you'd use `turbo run build` to prepare each of the artifacts through the entire tree.

> it doesn't automatically links the modules again

I'm surprised by this; and I'm wondering if you can provide more detail? This isn't how it's supposed to work...
